bolt-modules/boltlib/lib/puppet/functions/puppetize.rb

```diff
@@ -0,0 +1,49 @@
+# frozen_string_literal: true
+
+require 'fileutils'
+
+Puppet::Functions.create_function(:puppetize) do
+  dispatch :puppetize do
+    param 'TargetSpec', :targets
+  end
+
+  def puppetize(target_spec)
+    applicator = Puppet.lookup(:apply_executor) { nil }
+    executor = Puppet.lookup(:bolt_executor) { nil }
+    inventory = Puppet.lookup(:bolt_inventory) { nil }
+    unless applicator && executor && inventory && Puppet.features.bolt?
+      raise Puppet::ParseErrorWithIssue.from_issue_and_stack(
+        Puppet::Pops::Issues::TASK_MISSING_BOLT, action: _('puppetize')
+      )
+    end
+
+    targets = inventory.get_targets(target_spec)
+
+    # Ensure Puppet is installed
+    versions = call_function(:run_task, 'puppet_agent::version', targets)
```

Comment:
If done in the `facts` plan, we can reuse this task to determine whether to sync over custom facts. We would need to rewrite (or have the applicator provide) a custom function to collect all the module code into a temp moduledir.

I also think we'd want https://github.com/puppetlabs/bolt/pull/539 along with my suggestion to be able to upload directly to a temporary directory. Without that it's much harder to handle the platform variations when trying to upload files and use them.

The fact task would also need an option to collect custom facts, with some way to pass in paths for Facter `--custom-dir` and `--external-dir`.

Reply:
This function would be re-usable with Orchestrator if we pulled the custom_facts task into a module. I've initially avoided that because it's so closely tied to how we send the custom facts data and that may change.